File: frontend/src/lib/portfolio.ts
/**
 * Portfolio overview hook + types — powers the Home page (Task 24).
 *
 * The backend `/api/portfolio/overview` already aggregates stats and
 * pre-shapes per-stock rows from each stock's latest live snapshot,
 * so the Home page only needs to filter/sort and render.
 */

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "@/lib/api";
import { stocksKey } from "@/lib/stocks";

export interface SparklinePoint {
  created_at: string;
  general_grade: string | null;
}

export interface ScenarioFairValue {
  fair_value: number | null;
  upside_pct: number | null;
}

export interface PortfolioStockRow {
  symbol: string;
  currency: string;
  category: string | null;
  notes: string | null;
  last_updated: string | null;
  general_grade: string | null;
  sub_grades: Record<string, string | null>;
  average_fair_value: string | null;
  current_price: string | null;
  upside_pct: string | null;
  /** Per-scenario fair values from the latest fair-values job (null if none). */
  fair_values: Record<string, ScenarioFairValue> | null;
  sparkline: SparklinePoint[];
}

export interface PortfolioStats {
  total_stocks: number;
  average_general_grade: string | null;
  undervalued_count: number;
  overvalued_count: number;
}

export interface PortfolioOverview {
  stats: PortfolioStats;
  stocks: PortfolioStockRow[];
}

export const PORTFOLIO_QUERY_KEY = ["portfolio", "overview"] as const;

export function usePortfolioOverview() {
  return useQuery({
    queryKey: PORTFOLIO_QUERY_KEY,
    queryFn: () => api<PortfolioOverview>("/api/portfolio/overview"),
  });
}

/** Set (or clear, with null) a stock's category from the portfolio table. */
export function useSetCategory() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ symbol, category }: { symbol: string; category: string | null }) =>
      api<unknown>(`/api/stocks/${encodeURIComponent(symbol)}`, {
        method: "PATCH",
        body: { category },
      }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: PORTFOLIO_QUERY_KEY });
      queryClient.invalidateQueries({ queryKey: stocksKey });
    },
  });
}

/** Delete a ticker and all its data. */
export function useDeleteTicker() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (symbol: string) =>
      api<void>(`/api/stocks/${encodeURIComponent(symbol)}`, { method: "DELETE" }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: PORTFOLIO_QUERY_KEY });
      queryClient.invalidateQueries({ queryKey: stocksKey });
    },
  });
}

File: backend/app/schemas/portfolio.py
"""Pydantic schemas for the portfolio overview endpoint (Home page)."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel


class SparklinePoint(BaseModel):
    created_at: datetime
    general_grade: Decimal | None = None


class PortfolioStockRow(BaseModel):
    symbol: str
    currency: str
    category: str | None = None
    notes: str | None = None
    last_updated: datetime | None = None
    general_grade: Decimal | None = None
    sub_grades: dict[str, Decimal | None] = {}
    average_fair_value: Decimal | None = None
    current_price: Decimal | None = None
    upside_pct: Decimal | None = None
    # Per-scenario fair values from the latest fair-values job:
    # { scenario_key: {"fair_value": float|None, "upside_pct": float|None} }.
    fair_values: dict[str, Any] | None = None
    sparkline: list[SparklinePoint] = []


class PortfolioStats(BaseModel):
    total_stocks: int
    average_general_grade: Decimal | None = None
    undervalued_count: int
    overvalued_count: int


class PortfolioOverview(BaseModel):
    stats: PortfolioStats
    stocks: list[PortfolioStockRow]
